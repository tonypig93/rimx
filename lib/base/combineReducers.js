export function combineReducers(reducers) {
    return (state, action) => {
        const type = action.type;
        if (typeof reducers[type] === 'function') {
            return reducers[type](state, action);
        }
        console.warn('reducer is not a function');
        return state;
    };
}
