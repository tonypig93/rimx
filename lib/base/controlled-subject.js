/**
 * @class ControlledSubject
 */
import { Observable } from 'rxjs/Observable';
import { Subject } from 'rxjs/Subject';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/distinctUntilChanged';
import 'rxjs/add/operator/takeUntil';
import { compareFn } from './utils';
export class ControlledSubject {
    constructor(path, scopeId, root) {
        this.unsubscribe$ = new Subject();
        this.dispatch = (action) => {
            const reducer = this.root.SCOPE[this.path].reducer;
            this.next(state => reducer(state, action));
        };
        this.path = path;
        this.pluckPath = path.split('.');
        this.root = root;
        this.scopeId = scopeId;
        this.closed = false;
        this.stateObservable = root.store
            .asObservable()
            .map((rootState) => rootState.getIn(this.pluckPath));
    }
    /**
     *
     *
     * @param {*} observer
     * @returns
     * @memberof ControlledSubject
     */
    subscribe(observer, key, mapper) {
        const root = this.root;
        // root.takeSnapshot();
        let observable = this.stateObservable;
        if (key) {
            observable = observable.map((d) => d.getIn(key)).distinctUntilChanged(compareFn);
        }
        else {
            observable = observable.distinctUntilChanged(compareFn);
        }
        if (mapper) {
            observable = mapper(observable);
        }
        const subscription = observable.takeUntil(this.unsubscribe$).subscribe(observer);
        return subscription;
    }
    next(input) {
        if (this.closed) {
            return;
        }
        let nextState;
        const root = this.root;
        if (typeof input === 'function') {
            const snapshot = root._getSnapshot(this.pluckPath); // eslint-disable-line
            nextState = input(snapshot);
        }
        else {
            nextState = input;
        }
        if (nextState instanceof Observable) {
            nextState.subscribe(_data => {
                root.updateState(this.path, _data);
            });
        }
        else {
            root.updateState(this.path, nextState);
        }
    }
    snapshot() {
        return this.root._getSnapshot(this.pluckPath); // eslint-disable-line
    }
    destroy() {
        this.unsubscribe$.next();
        this.unsubscribe$.complete();
        this.closed = true;
        this.root = null;
        this.stateObservable = null;
    }
}
