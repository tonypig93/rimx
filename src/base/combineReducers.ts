import { Reducer, Action } from './types';

interface ReducersSource {
  [type: string]: Reducer;
}

export function combineReducers(reducers: ReducersSource) {
  return (state, action: Action) => {
    const type = action.type;
    if (typeof reducers[type] === 'function') {
      return reducers[type](state, action);
    }
    throw new Error('Expected the reducer to be a function');
  };
}
