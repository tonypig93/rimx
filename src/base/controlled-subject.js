/**
 * @class ControlledSubject
 */
import { compareFn } from './utils';

export default class ControlledSubject {
    constructor(path, scopeId, root) {
      this.path = path;
      this.pluckPath = path.split('.');
      this.root = root;
      this.scopeId = scopeId;
      this.streamControl = new Subject();
      this.closed = false;
      this.stateObservable = root.store
        .asObservable()
        .map((rootState) => rootState.getIn(this.pluckPath));
      // .finally(() => console.log('your watch is over'));
    }
    /**
     *
     *
     * @param {*} observer
     * @returns
     * @memberof ControlledSubject
     */
    subscribe(observer, key, mapper) {// eslint-disable-line
      const root = this.root;
      // root.takeSnapshot();
      let observable = this.stateObservable;
      if (key) {
        observable = observable.map((d) => d.getIn(key)).distinctUntilChanged(compareFn);
      } else {
        observable = observable.distinctUntilChanged(compareFn);
      }
      if (mapper) {
        observable = mapper(observable);
      }
      const subscription = observable.subscribe(observer);
      const observerScopeControl = root.findScopeObservers(this.scopeId);
      if (observerScopeControl) {
        observerScopeControl.observers.push(subscription);
        root.observers.set(root.observers.findIndex((ob) => {
          if (!ob) {
            throw new Error(`cannot find the observer list of scope ${this.scopeId}`);
          }
          return ob.$scopeId === this.scopeId;
        }), observerScopeControl);
      } else {
        throw new Error('Unknow error');
      }
      const unsubscribe = subscription.unsubscribe;
      subscription.unsubscribe = function _unsubscribe() {
        // const index = observerScopeControl.observers.findIndex((item) => item === subscription);
        // todo: delete item in array
        unsubscribe.call(subscription);
      };
      return subscription;
    }
    next(input) {
      if (this.closed) {
        return;
      }
      let newData;
      const root = this.root;
      if (typeof input === 'function') {
        const snapshot = root._getSnapshot(this.pluckPath);// eslint-disable-line
        newData = input(snapshot);
      } else {
        newData = input;
      }
      root.updateScope(this.path, newData);
    }
    snapshot() {
      return this.root._getSnapshot(this.pluckPath);// eslint-disable-line
    }
    destroy() {
      const root = this.root;
      const observerScopeControl = root.findScopeObservers(this.scopeId);
      this.closed = true;
  
      if (observerScopeControl) {
        observerScopeControl.observers.forEach((ob) => ob.unsubscribe());
      }
      root.deleteScope(this.path);
      const index = root.observers.findIndex((ob) => {
        if (!ob) {
          throw new Error(`cannot find the observer list of scope ${this.scopeId}`);
        }
        return ob.$scopeId === this.scopeId;
      });
      root.observers = root.observers.delete(index);
    }
  }
  