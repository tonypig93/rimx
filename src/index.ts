export { connect, RxStore } from './reactBinding';
export { combineReducers } from './base/combineReducers';
