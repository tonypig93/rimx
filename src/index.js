import React from 'react';
import { isPlainObject, toCamelcase } from './utils';
import { RxStoreFactory } from './base/factory';


export const RxStore = new RxStoreFactory();


/**
 * 宿主组件：创建scope的组件；连接组件：连接到其他宿主组件的scope的组件。
 * scopeName是字符串时，表示要创建新的scope，此时initState有效，connectScopes表示需要连接的scope
 * scopeName是对象时，表示要创建连接组件，scopeName的key为要连接的scopeName，value为函数时，函数需要从scope snapshot中返回需要的值（mapStateToProps），prop key为key。
 * value为对象时，{ propName: string; map: (data) => data }
 *
 * @export
 * @param {any} scopeName
 * @param {any} initState
 * @param {any} connectScopes
 * @returns
 */
export function connect(scopeName, initState, connectScopes) {
  return function wrap(WrapComponent) {
    return class WrappedComponent extends React.PureComponent {
      constructor() {
        super();
        this.subject = null;
        this.state = {};
        this.listeners = [];
        this.isConnected = false;
        this.stateToPropsNames = [];
        if (typeof scopeName === 'string') {
          this.createScope();
        }
        if (isPlainObject(scopeName) || isPlainObject(connectScopes)) {
          this.connectOptions = isPlainObject(scopeName) ? scopeName : connectScopes;
          this.connectScope(this.connectOptions);
          this.isConnected = true;
        }
      }
      componentWillMount() {
        this.mapStateToProps(this.subject);
      }
      get debugID() {
        return this._reactInternalInstance._debugID; //eslint-disable-line
      }
      createScope() { //eslint-disable-line
        RxStore.injectScope(scopeName, initState);
        this.subject = this.bindListener(RxStore.getStateSubject(scopeName));
        this.isScopeHoster = true;
      }
      connectScope(scopes) {
        if (this.subject) {
          this.subject = {
            [scopeName]: this.subject,
          };
        } else {
          this.subject = {};
        }
        Object.keys(scopes).filter((key) => key !== scopeName).forEach((key) => {
          const _subject = RxStore.getStateSubject(key);//eslint-disable-line
          this.subject[key] = this.bindListener(_subject);
        });
      }
      componentWillUnmount() {
        this.listeners.forEach((listener) => {
          listener.unsubscribe();
        });
      }
      bindListener(subject) {
        const bindedSubject = subject;
        bindedSubject.listen = (observer, k, mapper) => {
          const subscription = subject.subscribe(observer, k, mapper);
          this.listeners.push(subscription);
          return subscription;
        };
        return bindedSubject;
      }
      mapStateToProps(subject) {
        if (this.isConnected) {
          Object.keys(this.connectOptions).forEach((key) => {
            const mapProps = this.connectOptions[key];
            const subj = subject[key];
            if (typeof mapProps === 'string') {
              this.listenState(subj, toCamelcase(mapProps));
            } else if (isPlainObject(mapProps)) {
              this.listenState(subj, toCamelcase(mapProps.propName), mapProps.path);
            } else if (Array.isArray(mapProps)) {
              mapProps.forEach((item) => {
                if (typeof item === 'string') {
                  this.listenState(subj, toCamelcase(item));
                } else if (isPlainObject(item)) {
                  this.listenState(subj, toCamelcase(item.propName), item.path);
                }
              });
            }
          });
        }
      }
      listenState(subject, name, path = [name]) {
        this.stateToPropsNames.push(name);
        subject.listen((d) => {
          this.setState({
            [name]: d,
          });
        }, path);
      }
      getProps() {
        const props = {};
        this.stateToPropsNames.forEach((name) => {
          props[name] = this.state[name];
        });
        return props;
      }
      getSubject(scopeName) {
        
      }
      render() {
        const valueProps = this.getProps();
        return (
          <WrapComponent
            {...this.props}
            {...valueProps}
            subject={this.subject} />
        );
      }
    };
  };
}
