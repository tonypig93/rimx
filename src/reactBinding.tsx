import * as React from 'react';
import { Subscription } from 'rxjs/Subscription';
import { Observable } from 'rxjs/Observable';

import { normalizePath, isPlainObject, toCamelcase } from './base/utils';
import { RxStoreFactory } from './base/factory';
import { ScopeController } from './base/scopeController';
import { Reducer, Action } from './base/types';

interface ReactScopeController extends ScopeController {
  listen?: (
    key: string[],
    force: boolean,
  ) => {
    do: (observer) => Subscription;
    pipe: (
      mapper: (ob: Observable<any>) => Observable<any>
    ) => {
      do: (observer) => Subscription;
    };
  };
}

interface Options {
  scope?: string;
  scopeName?: string;
  initState?: any;
  connectScopes?: {
    [scopeName: string]: any;
  };
  reducer?: Reducer;
  cache?: boolean;
  log?: boolean;
}

const RxStore = new RxStoreFactory();

function connect(options: Options) {
  if (options.scope) {
    options.scopeName = options.scope;
  }
  return function wrap(WrappedComponent) {
    return class ConnectedComponent extends React.Component<any, any> {
      controllerSet: { [key: string]: ReactScopeController } = {};
      state = {};
      isConnected = false;
      isScopeRoot = false;
      stateToPropsNames: string[] = [];
      connectOptions: any;
      constructor(props, context) {
        super(props, context);
        if (typeof options.scopeName === 'string') {
          this.createScope(
            options.scopeName,
            options.reducer,
            options.cache,
            options.log
          );
        }
        if (isPlainObject(options.connectScopes)) {
          this.connectOptions = options.connectScopes;
          this.connectScope(this.connectOptions);
          this.isConnected = true;
        }
      }

      componentWillMount() {
        this.mapStateToProps(this.controllerSet);
      }

      shouldComponentUpdate(nextProps, nextState) {
        if (nextProps !== this.props || nextState !== this.state) {
          return true;
        }
        return false;
      }

      componentWillUnmount() {
        Object.keys(this.controllerSet).forEach(key => {
          this.controllerSet[key].destroy();
        });

        this.controllerSet = null;
      }

      createScope(
        name: string,
        reducer: Reducer,
        cache: boolean,
        log: boolean
      ) {
        this.isScopeRoot = true;
        RxStore.injectScope(name, options.initState, reducer, cache, log);
        this.controllerSet[name] = this.bindListener(RxStore.getScopeController(name));
      }

      connectScope(scopes) {
        Object.keys(scopes)
          .filter(key => key !== options.scopeName)
          .forEach(key => {
            const scopeController = RxStore.getScopeController(key);
            this.controllerSet[key] = this.bindListener(scopeController);
          });
      }

      bindListener(controller: ScopeController) {
        const bindedController: ReactScopeController = controller;
        bindedController.listen = (key, force) => {
          let _mapper;

          const _do = observer => {
            const subscription = controller.subscribe(observer, key, force, _mapper);
            return subscription;
          };

          function pipe(mapper) {
            _mapper = mapper;
            return {
              do: _do
            };
          }

          return {
            do: _do,
            pipe
          };
        };
        return bindedController;
      }

      mapStateToProps(controllerSet) {
        if (this.isConnected) {
          Object.keys(this.connectOptions).forEach(key => {
            const mapProps = this.connectOptions[key];
            const controller = controllerSet[key];
            if (typeof mapProps === 'string') {
              this.listenState(controller, toCamelcase(mapProps));
            } else if (isPlainObject(mapProps)) {
              this.listenState(
                controller,
                toCamelcase(mapProps.propName),
                mapProps.path,
                mapProps.selector,
              );
            } else if (Array.isArray(mapProps)) {
              mapProps.forEach(item => {
                if (typeof item === 'string') {
                  this.listenState(controller, toCamelcase(item));
                } else if (isPlainObject(item)) {
                  this.listenState(controller, toCamelcase(item.propName), item.path, item.selector);
                }
              });
            }
          });
        }
      }

      listenState(subject: ReactScopeController, name: string, path = [name], selector?) {
        this.stateToPropsNames.push(name);
        subject
          .listen(normalizePath(path), false)
          .pipe(ob => selector ? ob.map(selector) : ob)
          .do(d => {
            this.setState({
              [name]: d
            });
          });
      }

      getPropsInState() {
        const props = {};
        this.stateToPropsNames.forEach(name => {
          props[name] = this.state[name];
        });
        return props;
      }

      getInjectProps() {
        // subject is deprecated, use controller(s) instead
        const controllerKeys = Object.keys(this.controllerSet);
        let props;
        if (controllerKeys.length === 1) {
          const controller = this.controllerSet[controllerKeys[0]];
          props = {
            listen: controller.listen,
            dispatch: controller.dispatch,
            subject: controller,
            controller,
          };
        } else {
          props = {
            subject: this.controllerSet,
            controllers: this.controllerSet,
          };
        }
        return props;
      }

      render() {
        return (
          <WrappedComponent
            {...this.getPropsInState()}
            {...this.getInjectProps()}
            {...this.props}
          />
        );
      }
    };
  };
}

export {
  connect,
  RxStore,
}
