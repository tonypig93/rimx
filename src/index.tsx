import * as React from 'react';
import { Subscription, Observable } from 'rxjs';

import { isPlainObject, toCamelcase, normalizePath } from './utils';
import { RxStoreFactory } from './base/factory';
import { ControlledSubject } from './base/controlled-subject';
import { Reducer, Action } from './base/types';
export { combineReducers } from './base/combineReducers';

interface ReactSubject extends ControlledSubject {
  listen?: (key: string[]) => {
    do: (observer) => Subscription,
    pipe: (ob: Observable<any>) => {
      do: (observer) => Subscription,
    },
  }
}

export const RxStore = new RxStoreFactory();
/**
 * 宿主组件：创建scope的组件；连接组件：连接到其他宿主组件的scope的组件。
 * scopeName是字符串时，表示要创建新的scope，此时initState有效，connectScopes表示需要连接的scope
 * scopeName是对象时，表示要创建连接组件，scopeName的key为要连接的scopeName，value为函数时，函数需要从scope snapshot中返回需要的值（mapStateToProps），prop key为key。
 * value为对象时，{ propName: string; map: (data) => data }
 *
 * @export
 * @param {any} scopeName
 * @param {any} initState
 * @param {any} connectScopes
 * @returns
 */
export function connect(scopeName: any, initState, connectScopes, reducer: Reducer) {
  return function wrap(WrapComponent) {
    return class WrappedComponent extends React.PureComponent<any, any> {
      subjectMap: { [key: string]: ReactSubject } = {};
      state = {};
      isConnected = false;
      isScopeRoot = false;
      stateToPropsNames: string[] = [];
      connectOptions: any;
      constructor(p, s) {
        super(p, s);
        if (typeof scopeName === 'string') {
          if (typeof connectScopes === 'function') {
            reducer = connectScopes;
          }
          this.createScope(scopeName, reducer);
        }
        if (isPlainObject(scopeName) || isPlainObject(connectScopes)) {
          this.connectOptions = isPlainObject(scopeName) ? scopeName : connectScopes;
          this.connectScope(this.connectOptions);
          this.isConnected = true;
        }
      }
      componentWillMount() {
        this.mapStateToProps(this.subjectMap);
      }
      componentWillUnmount() {
        Object.keys(this.subjectMap).forEach(key => {
          this.subjectMap[key].destroy();
        });

        this.subjectMap = null;
      }
      createScope(name: string, reducer: Reducer) {
        this.isScopeRoot = true;
        RxStore.injectScope(name, initState, reducer);
        this.subjectMap[name] = this.bindListener(RxStore.getStateSubject(name));
      }
      connectScope(scopes) {
        Object.keys(scopes).filter((key) => key !== scopeName).forEach((key) => {
          const _subject = RxStore.getStateSubject(key);
          this.subjectMap[key] = this.bindListener(_subject)
        });
      }

      bindListener(subject: ControlledSubject) {
        const bindedSubject: ReactSubject = subject;
        bindedSubject.listen = (key) => {
          let _mapper;
          
          const _do = (observer) => {
            const subscription = subject.subscribe(observer, key, _mapper);
            return subscription;
          }
          
          function pipe(mapper) {
            _mapper = mapper;
            return {
              do: _do,
            };
          }

          return {
            do: _do,
            pipe,
          }
        }
        return bindedSubject;
      }
      mapStateToProps(subject) {
        if (this.isConnected) {
          Object.keys(this.connectOptions).forEach((key) => {
            const mapProps = this.connectOptions[key];
            const subj = subject[key];
            if (typeof mapProps === 'string') {
              this.listenState(subj, toCamelcase(mapProps));
            } else if (isPlainObject(mapProps)) {
              this.listenState(subj, toCamelcase(mapProps.propName), mapProps.path);
            } else if (Array.isArray(mapProps)) {
              mapProps.forEach((item) => {
                if (typeof item === 'string') {
                  this.listenState(subj, toCamelcase(item));
                } else if (isPlainObject(item)) {
                  this.listenState(subj, toCamelcase(item.propName), item.path);
                }
              });
            }
          });
        }
      }
      listenState(subject: ReactSubject, name, path = [name]) {
        this.stateToPropsNames.push(name);
        subject
          .listen(normalizePath(path))
          .do(d => {
            this.setState({
              [name]: d,
            });
          });
      }
      getPropsInState() {
        const props = {};
        this.stateToPropsNames.forEach((name) => {
          props[name] = this.state[name];
        });
        return props;
      }
      getInjectProps() {
        const subjectsKey = Object.keys(this.subjectMap);
        let props;
        if (subjectsKey.length === 1) {
          const subject = this.subjectMap[subjectsKey[0]];
          props = {
            listen: subject.listen,
            dispatch: subject.dispatch,
            subject,
          };
        } else {
          props = {
            subject: this.subjectMap,
          };
        }
        return props;
      }
      render() {
        return (
          <WrapComponent
            {...this.getPropsInState()}
            {...this.getInjectProps()}
            {...this.props}
          />
        );
      }
    };
  };
}
