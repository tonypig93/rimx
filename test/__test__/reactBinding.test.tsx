import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { connect } from '../../src/reactBinding';
import { combineReducers } from '../../src/base/combineReducers';

let Root$;

class Root extends React.Component<any, any> {
  handleChangeState = () => {
    this.props.dispatch({ type: 'changeName', payload: 'tom' });
  }
  render() {
    return (
      <div onClick={this.handleChangeState}>{this.props.name}</div>
    )
  }
}

class Another extends React.Component<any, any> {
  render() {
    return (
      <div>{this.props.name}</div>
    )
  }
}

beforeEach(() => {
  Root$ = connect({
    scope: 'test',
    initState: {
      name: 'tony',
      list: [],
    },
    connectScopes: {
      test: 'name',
    },
    reducer,
  })(Root);
})

function reducer1() {
  return {
    name: 'tom',
  };
}

const reducer = combineReducers({
  changeName: reducer1,
});

describe('React bindings', () => {
  test('create a scope with a component', () => {
    expect(Root$.name).toBe('ConnectedComponent');

    const component = renderer.create(
      <Root$></Root$>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    tree.props.onClick();

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })

  test('make a component link to the scope', () => {
    const Another$ = connect({
      connectScopes: {
        test: 'name',
      },
    })(Another);

    const component_r = renderer.create(
      <Root$></Root$>
    );

    const component_a = renderer.create(
      <Another$></Another$>
    );

    let tree_r = component_r.toJSON();
    expect(tree_r).toMatchSnapshot();

    let tree_a = component_a.toJSON();
    expect(tree_a).toMatchSnapshot();

    tree_r.props.onClick();

    tree_r = component_r.toJSON();
    expect(tree_r).toMatchSnapshot();

    tree_a = component_a.toJSON();
    expect(tree_a).toMatchSnapshot();

  });
})
