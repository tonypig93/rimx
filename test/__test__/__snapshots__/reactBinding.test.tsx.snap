// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React bindings create a scope with a component 1`] = `
<div
  onClick={[Function]}
>
  tony
</div>
`;

exports[`React bindings create a scope with a component 2`] = `
<div
  onClick={[Function]}
>
  tom
</div>
`;

exports[`React bindings make a component link to the scope 1`] = `
<div
  onClick={[Function]}
>
  tony
</div>
`;

exports[`React bindings make a component link to the scope 2`] = `
<div>
  tony
</div>
`;

exports[`React bindings make a component link to the scope 3`] = `
<div
  onClick={[Function]}
>
  tom
</div>
`;

exports[`React bindings make a component link to the scope 4`] = `
<div>
  tom
</div>
`;
